components/client/typescript/src/server.ts

```diff
@@ -0,0 +1,187 @@
+
+const IfThisThenNothingSchema = Type.Union([
+  Type.Composite([
+    BitcoinIfThisOptionsSchema,
+    Type.Object({
+      if_this: BitcoinIfThisSchema,
+    }),
+  ]),
+  Type.Composite([
+    StacksIfThisOptionsSchema,
+    Type.Object({
+      if_this: StacksIfThisSchema,
+    }),
+  ]),
+]);
+const ServerPredicateSchema = Type.Composite([
+  PredicateHeaderSchema,
+  Type.Object({
+    networks: Type.Union([
+      Type.Object({
+        mainnet: IfThisThenNothingSchema,
+      }),
+      Type.Object({
+        testnet: IfThisThenNothingSchema,
+      }),
+    ]),
+  }),
+]);
+/** Chainhook predicates registerable by the local event server */
+export type ServerPredicate = Static<typeof ServerPredicateSchema>;
+
+/**
+ * Build the Chainhook Fastify event server.
+ * @param serverOpts - Server options
+ * @param chainhookOpts - Chainhook node options
+ * @param predicates - Predicates to register
+ * @param callback - Event callback function
+ * @returns Fastify instance
+ */
+export async function buildServer(
+  serverOpts: ServerOptions,
+  chainhookOpts: ChainhookNodeOptions,
+  predicates: [ServerPredicate],
+  callback: OnEventCallback
+) {
+  const base_path = `http://${chainhookOpts.hostname}:${chainhookOpts.port}`;
+
+  async function waitForNode(this: FastifyInstance) {
+    logger.info(`EventServer connecting to chainhook node at ${base_path}...`);
+    while (true) {
+      try {
+        await request(`${base_path}/ping`, { method: 'GET', throwOnError: true });
+        break;
+      } catch (error) {
+        logger.error(error, 'Chainhook node not available, retrying...');
+        await timeout(1000);
+      }
+    }
+  }
+
+  async function registerPredicates(this: FastifyInstance) {
+    logger.info(predicates, `EventServer registering predicates at ${base_path}...`);
+    for (const predicate of predicates) {
+      const thenThat: ThenThatHttpPost = {
+        http_post: {
+          url: `${serverOpts.external_base_url}/chainhook/${predicate.uuid}`,
+          authorization_header: `Bearer ${serverOpts.auth_token}`,
+        },
+      };
+      try {
+        const body = predicate as Predicate;
+        if ('mainnet' in body.networks) body.networks.mainnet.then_that = thenThat;
+        if ('testnet' in body.networks) body.networks.testnet.then_that = thenThat;
+        await request(`${base_path}/v1/chainhooks`, {
+          method: 'POST',
+          body: JSON.stringify(body),
+          headers: { 'content-type': 'application/json' },
+          throwOnError: true,
+        });
+        logger.info(`EventServer registered '${predicate.name}' predicate (${predicate.uuid})`);
+      } catch (error) {
+        logger.error(error, `EventServer unable to register predicate`);
+      }
+    }
+  }
+
+  async function removePredicates(this: FastifyInstance) {
+    logger.info(`EventServer closing predicates at ${base_path}...`);
+    for (const predicate of predicates) {
```

Comment:
If you wanted to this to complete ASAP (e.g. during shutdown where grace periods may only be 1000ms), you could init all these network requests at once and put them in an array of promises, then use something like `await Promise.allSettled(removalPromises)`.

(Just a suggestion, not that important seeming)